File: MobileAcceleratorExample/AICVC.m
//
//  AICVC.m
//  MobileAcceleratorExample
//
//  Created by Brian Salomon on 8/8/16.
//  Copyright © 2016 Akamai. All rights reserved.
//

#import "AICVC.h"
#import "AppDelegate.h"
#import <AkaMap/AkaMap.h>
#import <AkaCommon/AkaCommon.h>

@interface AICVC ()

@property (nonatomic, strong) UIImage *image;
@property (nonatomic, strong) NSMutableData *responseData;
@property (nonatomic, assign) NSInteger responseStatusCode;
@property (nonatomic) NSTimeInterval startTime;
@property (nonatomic) NSTimeInterval loadTimeElapsed;

@end

@implementation AICVC

- (void)viewDidLoad
{
    [super viewDidLoad];

	self.labelURL.text = [NSString stringWithFormat:@"URL: %@", [self.url absoluteString]];

	[self beginDownload];
}

- (void)didReceiveMemoryWarning
{
    [super didReceiveMemoryWarning];
}

- (void)beginDownload
{
	self.labelResponseCode.text = @"Response Code: ---";
	self.labelContentSize.text = @"Content Size: --- bytes";
	self.labelStatus.text = @"Status: reloading";

	self.startTime = [NSDate timeIntervalSinceReferenceDate];

	// Create config using Mobile Accelerator SDK
	NSURLSessionConfiguration *sessionConfig = [NSURLSessionConfiguration defaultSessionConfiguration];
	[[AkaCommon shared] interceptSessionsWithConfiguration:sessionConfig];

	NSURLSession *session = [NSURLSession sessionWithConfiguration:sessionConfig delegate:self delegateQueue:nil];

	// Create a request that ignores the system cache. This allows
	// for reloading the image under different network conditions.
	NSMutableURLRequest *request = [NSMutableURLRequest requestWithURL:self.url];
	request.cachePolicy = NSURLRequestReloadIgnoringLocalCacheData;

	NSURLSessionDataTask *dataTask = [session dataTaskWithRequest:request];
	[dataTask resume];
	[session finishTasksAndInvalidate];
}

// run on main thread due to UIKit reference
- (void)updateImage
{
    self.image = [UIImage imageWithData:self.responseData];
    [self.imageView setImage:self.image];
}

- (IBAction)tappedReload:(id)sender
{
	self.responseData = nil;
	[self updateImage];

	[self beginDownload];
}

- (IBAction)tappedChangeURL:(id)sender
{
	UIAlertController *alert = [UIAlertController alertControllerWithTitle:@"Change URL"
																   message:nil
															preferredStyle:UIAlertControllerStyleAlert];

	[alert addAction:[UIAlertAction actionWithTitle:@"Done" style:UIAlertActionStyleDefault handler:^(UIAlertAction *action){

		// set and perform an immediate download of the new URL
		UITextField *textField = alert.textFields[0];
		self.url = [NSURL URLWithString:textField.text];
		self.labelURL.text = [NSString stringWithFormat:@"URL: %@", [self.url absoluteString]];
		[self beginDownload];

	}]];
	[alert addAction:[UIAlertAction actionWithTitle:@"Cancel" style:UIAlertActionStyleCancel handler:nil]];
	[alert addTextFieldWithConfigurationHandler:^(UITextField *textField) {
		textField.text = [self.url absoluteString];
	}];
	[self presentViewController:alert animated:YES completion:nil];
}

#pragma mark NSURLSessionDelegate

- (void)URLSession:(NSURLSession *)session didBecomeInvalidWithError:(NSError *)error
{
    NSLog(@"NSURLSessionDelegate error: %@", error);
}

#pragma mark NSURLSessionDataDelegate

- (void)URLSession:(NSURLSession *)session dataTask:(NSURLSessionDataTask *)dataTask didReceiveResponse:(NSURLResponse *)response completionHandler:(void (^)(NSURLSessionResponseDisposition))completionHandler
{
    NSHTTPURLResponse *httpResponse = (NSHTTPURLResponse*)response;
    if (httpResponse) {
        self.responseStatusCode = httpResponse.statusCode;
    }

    self.responseData = [[NSMutableData alloc] init];
    completionHandler(NSURLSessionResponseAllow);
}

- (void)URLSession:(NSURLSession *)session dataTask:(NSURLSessionDataTask *)dataTask didReceiveData:(NSData *)data
{
    [self.responseData appendData:data];
}

/**
 * Disable response from being cached.
 */
- (void)URLSession:(NSURLSession *)session
		  dataTask:(NSURLSessionDataTask *)dataTask
 willCacheResponse:(NSCachedURLResponse *)proposedResponse
 completionHandler:(void (^)(NSCachedURLResponse *cachedResponse))completionHandler
{
	completionHandler(NULL);
}

#pragma mark NSURLSessionTaskDelegate

- (void)URLSession:(NSURLSession *)session task:(NSURLSessionTask *)task didCompleteWithError:(NSError *)error
{
    if (error) {
        NSLog(@"NSURLSessionTaskDelegate error: %@", [error localizedDescription]);
        dispatch_async(dispatch_get_main_queue(), ^{
            self.labelStatus.text = [NSString stringWithFormat:@"Status: %@", [error localizedDescription]];
        });
        return;
    }

	NSDate *dateNow = [NSDate date];
    self.loadTimeElapsed = [dateNow timeIntervalSinceReferenceDate] - self.startTime;

    dispatch_async(dispatch_get_main_queue(), ^{
		[self updateImage];

		NSDateFormatter *dateFormat = [NSDateFormatter new];
		[dateFormat setDateFormat:@"hh:mm:ss a"];
		NSString *dateString = [dateFormat stringFromDate:dateNow];

		self.labelResponseCode.text = [NSString stringWithFormat:@"Response Code: %ld", (long)self.responseStatusCode];
		self.labelContentSize.text = [NSString stringWithFormat:@"Content Size: %ld bytes", (unsigned long)self.responseData.length];
		self.labelStatus.text = [NSString stringWithFormat:@"Status: updated %@\nLoad Time: %ld ms",
								 dateString, (long)(self.loadTimeElapsed * 1000.0)];
    });
}

@end

File: MobileAcceleratorExampleSwift/AICVC.swift
//
//  AICVC.swift
//  MobileAcceleratorExampleSwift
//
//  Created by Waazim Reza on 09/08/2016.
//  Copyright © 2016 akamai. All rights reserved.
//

import Foundation
import AkaCommon

final class AICVC: UIViewController, URLSessionDelegate,URLSessionDataDelegate,URLSessionTaskDelegate {

	var url: URL?

	@IBOutlet weak var labelContentSize: UILabel!
	@IBOutlet weak var labelResponseCode: UILabel!
	@IBOutlet weak var labelURL: UILabel!
	@IBOutlet weak var labelStatus: UILabel!

	@IBOutlet weak var imageView: UIImageView!

	@IBOutlet weak var buttonReload: UIButton!
	@IBOutlet weak var buttonChangeURL: UIButton!

	private var image: UIImage!
	private var responseData: NSMutableData?
	private var responseStatusCode: NSInteger!

	private var startTime: TimeInterval = 0.0
	private var loadTimeElapsed: TimeInterval = 0.0

	override func viewDidLoad() {
		super.viewDidLoad()

		labelURL.text = "URL: \(url?.absoluteString ?? "<no URL>")"
		beginDownload()
	}

	func beginDownload() {
		labelResponseCode.text = "Response Code: ---"
		labelContentSize.text = "Content Size: --- bytes"
		labelStatus.text = "Status: reloading"

		guard let url = url else {
			print("No URL set.")
			return
		}

		startTime = Date.timeIntervalSinceReferenceDate

		// Create config using Mobile Accelerator SDK
		let sessionConfig: URLSessionConfiguration = URLSessionConfiguration.default
		AkaCommon.shared().interceptSessions(with: sessionConfig)

		let session = URLSession.init(configuration: sessionConfig, delegate: self, delegateQueue: nil)

		// Create a request that ignores the system cache. This allows
		// for reloading the image under different network conditions.
		var request = URLRequest(url: url)
		request.cachePolicy = .reloadIgnoringLocalCacheData

		let dataTask = session.dataTask(with: request)
		dataTask.resume()
	}

	// run on main thread due to UIKit reference
	func updateImage() {
		if let responseData = responseData as Data? {
			image = UIImage.init(data: responseData)
			imageView.image = image
		}
	}

	@IBAction func tappedReload(_ sender: AnyObject) {
		responseData = nil
		beginDownload()
	}

	@IBAction func tappedChangeURL(_ sender: AnyObject) {
		let alert = UIAlertController(title: "Change URL" , message: "", preferredStyle: .alert)

		alert.addAction(UIAlertAction(title: "Done", style: .default) { UIAlertAction in
			if let textFields = alert.textFields, textFields.count > 0, let text = textFields[0].text {
				self.url = URL(string: text)
				self.labelURL.text = "URL: \(self.url?.absoluteString ?? "<no URL>")"
				self.beginDownload()
			}
		})

		alert.addAction(UIAlertAction(title: "Cancel", style: .cancel, handler: nil))
		alert.addTextField(configurationHandler: { (textField) in
			textField.text = self.url?.absoluteString
		})

		present(alert, animated: true,completion: nil)
	}

	// MARK: URLSessionDelegate
	func urlSession(_ session: URLSession, didBecomeInvalidWithError error: Error?) {
		print("URLSessionDelegate error: \(error.debugDescription)")
	}

	// MARK: URLSessionDataDelegate
	func urlSession(_ session: URLSession, dataTask: URLSessionDataTask, didReceive response: URLResponse, completionHandler: @escaping (URLSession.ResponseDisposition) -> Swift.Void) {
		if let httpResponse = response as? HTTPURLResponse {
			responseStatusCode = httpResponse.statusCode
		}
		responseData = NSMutableData()
		completionHandler(.allow)
	}

	func urlSession(_ session: URLSession, dataTask: URLSessionDataTask, didReceive data: Data) {
		if let responseData = responseData {
			responseData.append(data)
		}
	}

	func urlSession(_ session: URLSession, dataTask: URLSessionDataTask, willCacheResponse proposedResponse: CachedURLResponse, completionHandler: @escaping (CachedURLResponse?) -> Swift.Void) {
		completionHandler(nil)
	}

	// MARK:  URLSessionTaskDelegate
	func urlSession(_ session: URLSession, task: URLSessionTask, didCompleteWithError error: Error?) {
		guard error == nil else {
			DispatchQueue.main.async() {
				print("URLSessionTaskDelegate error: \(error.debugDescription)")
				self.labelStatus.text = "Status: \(error.debugDescription)"
			}
			return
		}

		let dateNow = Date()
		loadTimeElapsed = dateNow.timeIntervalSinceReferenceDate - startTime

		DispatchQueue.main.async() {
			self.updateImage()

			let dateFormat = DateFormatter()
			dateFormat.dateFormat = "hh:mm:ss a"

			let dateString = dateFormat.string(from: dateNow)
			if let responseCode = self.responseStatusCode {
				self.labelResponseCode.text = "Response Code: \(responseCode)"
			} else {
				self.labelResponseCode.text = "Response Code: <none>"
			}
			if let responseData = self.responseData {
				self.labelContentSize.text = "Content Size:  \(responseData.length) bytes"
			} else {
				self.labelContentSize.text = "Content Size: <unknown>"
			}
			self.labelStatus.text = "Status: updated \(dateString)\nLoad Time: \(self.loadTimeElapsed * 1000.0) ms"
		}
	}
}
